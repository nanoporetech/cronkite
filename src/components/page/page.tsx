// tslint:disable: no-import-side-effect
import '@ionic/core';
import '@metrichor/ui-components';
import { Component, Event, EventEmitter, h, Host, Listen, Method, Prop, State, Watch } from '@stencil/core';
import Ajv from 'ajv';
import reportSchema from '../../cronkite.schema.json';
import { ComponentConfig, ReportDefinition, Stream } from '../../types/reportconfig.type';
import { debounce } from '../../utils';

@Component({
  assetsDirs: ['../../cronkite.schema.json'],
  styleUrl: 'page.scss',
  tag: 'cronk-page',
})
export class CronkPage {
  private ajv = new Ajv();

  /** broadcast to any listener (mostly datastreams) that the page has fully
   * rendered and payloads should be rebroadcast to ensure the lated data
   * is set on components
   */
  @Event({ bubbles: false, cancelable: true }) cronkPageReady!: EventEmitter<void>;

  @State() _pageConfig: ReportDefinition | undefined;
  @State() _schemaError = '';

  /** show or hide the configuration used to render the page */
  @Prop() showConfig = false;
  /** enable/disable validation ot the cronkite schema */
  @Prop({ reflect: true }) validationEnabled = true;
  /** Page configuration JSON */
  @Prop({ reflect: true, mutable: true }) pageConfig: ReportDefinition | string | undefined;

  @Watch('pageConfig')
  async watchHandler(newConfig: any) {
    const sanitized = this.coerceCronkiteConfig(newConfig);
    if ((await this.validateConfig(sanitized)) || !this.validationEnabled || newConfig === null) {
      this._schemaError = '';
      return;
    }
    this._schemaError = this.stringifyAjvError();
  }

  /** validate provided configuration JSON for a page */
  @Method()
  async validateConfig(configIn: any) {
    if (!configIn) return false;
    let isValid = false;
    const alarm = (this.validationEnabled && 'error') || 'warn';
    try {
      isValid = this.isValidConfig(configIn);
      if (!isValid) {
        console[alarm](
          `Cronkite schema validation ${alarm === 'error' ? alarm : 'warning'}: ${this.stringifyAjvError()}`,
        );
      }
    } catch (error) {
      console.error(error);
    }
    return isValid;
  }

  /** Load new page configuration JSON */
  @Method()
  async loadConfig(newConfig: any): Promise<void> {
    this.pageConfig = newConfig;
  }

  @Listen('componentsLoaded')
  async componentsLoadedHandler() {
    this.reloadStreams();
  }

  private reloadStreams = debounce(async () => {
    // console.debug('%cCRONK-PAGE::reloadStreams::RELOADING', 'color: mediumseagreen');
    this.cronkPageReady.emit();
    document.querySelector('cronk-datastreams')?.reload();
  }, 0);

  private coerceCronkiteConfig = (config: ReportDefinition | string): ReportDefinition | null => {
    let coercedConfig = config;
    if (typeof coercedConfig === 'string') {
      try {
        coercedConfig = JSON.parse(coercedConfig);
      } catch (error) {
        this._schemaError = error.message;
        return null;
      }
    }
    if (!coercedConfig) return null;
    return coercedConfig as ReportDefinition;
  };

  private stringifyAjvError() {
    return JSON.stringify(this.ajv.errors, null, 2);
  }

  private isValidConfig(configIn: any): boolean {
    if (!configIn) return false;
    return this.ajv.validate(reportSchema, configIn) as boolean;
  }

  private updatePageConfigState() {
    if (typeof this.pageConfig === 'string') {
      const coercedConfig = this.coerceCronkiteConfig(this.pageConfig);
      if (coercedConfig !== null) {
        this._pageConfig = coercedConfig;
        return;
      }
    }
    this._pageConfig = (this.pageConfig && (this.pageConfig as ReportDefinition)) || undefined;
  }

  componentWillUpdate() {
    this.updatePageConfigState();
  }

  componentWillLoad() {
    this.updatePageConfigState();
  }

  /* eslint-disable @stencil/render-returns-host */
  render() {
    const hasConfig = this._pageConfig !== undefined;
    const isValidConfig =
      (hasConfig && this.validationEnabled && this.isValidConfig(this._pageConfig)) || !this.validationEnabled;
    const classes = { 'empty-page-config': !hasConfig, 'valid-page-config': isValidConfig && this.validationEnabled };

    const { id, components, streams } = (isValidConfig && (this._pageConfig as ReportDefinition)) || {};

    const hasStreams = streams !== undefined && streams.length > 0;

    const streamsKey =
      (hasStreams &&
        (streams as Stream[])
          .reduce(
            (previousStreams: any, currentStream: any) => {
              return [...previousStreams, ...currentStream['@channels'].map((channel: any) => channel.channel)];
            },
            [id],
          )
          .join(':')) ||
      'no:streams';

    // console.debug('CRONK-PAGE::render', { hasConfig, isValidConfig, id, classes, components, streams, streamsKey });
    return <Host id={id} class={classes}>
      {/* RENDER COMPONENTS */}
      {(isValidConfig && (
        <cronk-page-components id={`${id}-components`} slot="components">
          {(components || []).map((compDef: ComponentConfig, componentIndex: number) => {
            const componentDefinition = compDef.layout ? compDef : { ...compDef, layout: {} };
            const uuid = `${id}-panel-${componentIndex}`;
            return (
              <cronk-page-panel
                slot={componentDefinition.layout.position}
                key={uuid}
                id={uuid}
                panelConfig={componentDefinition}
              />
            );
          })}
        </cronk-page-components>
      )) ||
        null}
      {/* RENDER STREAMS */}
      {(streams && streams.length && (
        <cronk-datastreams streams={streams} streamsID={streamsKey}></cronk-datastreams>
      )) ||
        null}
      {/* RENDER RAW CONFIG IF REQUESTED */}
      {(this.showConfig && (
        <pre>{isValidConfig ? JSON.stringify(this._pageConfig, null, 2) : 'No config provided'}</pre>
      )) ||
        null}
      {/* RENDER SCHEMA VALIDATION ERRORS */}
      {(this.validationEnabled && this._schemaError && <pre>{this._schemaError}</pre>) || null}
    </Host>
  }
}
