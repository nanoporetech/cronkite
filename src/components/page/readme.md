# cronk-page



<!-- Auto Generated Below -->


## Properties

| Property                  | Attribute                   | Description                                            | Type                         | Default     |
| ------------------------- | --------------------------- | ------------------------------------------------------ | ---------------------------- | ----------- |
| `disableSchemaValidation` | `disable-schema-validation` | enable/disable validation ot the cronkite schema       | `boolean`                    | `false`     |
| `pageConfig`              | `page-config`               | Page configuration JSON                                | `ReportDefinition \| string` | `undefined` |
| `showConfig`              | `show-config`               | show or hide the configuration used to render the page | `boolean`                    | `false`     |


## Events

| Event            | Description                                                                                                                                                      | Type                |
| ---------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------- |
| `cronkPageReady` | broadcast to any listener (mostly datastreams) that the page has fully rendered and payloads should be rebroadcast to ensure the lated data is set on components | `CustomEvent<void>` |


## Methods

### `loadConfig(newConfig: ReportDefinition | string) => Promise<void>`

Load new page configuration JSON

#### Returns

Type: `Promise<void>`



### `validateConfig(configIn: unknown) => Promise<boolean>`

validate provided configuration JSON for a page

#### Returns

Type: `Promise<boolean>`




## Dependencies

### Used by

 - [cronk-app](../../views/cronk-app)

### Depends on

- [cronk-page-components](../page-components)
- [cronk-page-panel](../page-panel)

### Graph
```mermaid
graph TD;
  cronk-page --> cronk-page-components
  cronk-page --> cronk-page-panel
  cronk-page-panel --> cronk-errormessage
  cronk-page-panel --> cronk-page-components
  cronk-page-panel --> cronk-page-panel
  cronk-app --> cronk-page
  style cronk-page fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
