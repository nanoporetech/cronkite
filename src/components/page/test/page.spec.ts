// tslint:disable: object-literal-sort-keys
import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { ReportDefinition } from '../../../types/reportconfig.type';
import { CronkPage } from '../page';

describe('cronk-page', () => {
  let rootInst: CronkPage;
  let rootEl: HTMLCronkPageElement;
  let page: SpecPage;
  const originalConsoleError = console.error as jest.Mock;

  beforeEach(async () => {
    (console.error as jest.Mock) = jest.fn();
    page = await newSpecPage({
      components: [CronkPage],
      html: '<cronk-page></cronk-page>',
    });
    rootInst = page.rootInstance;
    rootEl = page.root as HTMLCronkPageElement;
  });

  afterEach(async () => {
    ((console.error as jest.Mock) as jest.Mock).mockClear();
  });

  afterAll(() => {
    (console.error as jest.Mock) = originalConsoleError;
  });

  describe('sanity', () => {
    it('builds', () => {
      expect(rootInst).toBeTruthy();
    });

    it('has no Page config to start', async () => {
      page = await newSpecPage({
        components: [CronkPage],
        html: '<cronk-page></cronk-page>',
      });
      expect(page.rootInstance.pageConfig).toBeFalsy();
      expect(page.root).toEqualHtml('<cronk-page class="empty-page-config" validation-enabled></cronk-page>');
    });
  });

  describe('class methods', () => {
    it('isValidConfig() & stringifyAjvError()', async () => {
      page = await newSpecPage({
        components: [CronkPage],
        html: '<cronk-page></cronk-page>',
      });
      const cronkPage = page.rootInstance;
      expect(cronkPage.stringifyAjvError()).toBeUndefined();

      expect(cronkPage.isValidConfig({})).toEqual(false);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'id'`);

      expect(
        cronkPage.isValidConfig({
          id: 'test:page',
        }),
      ).toEqual(false);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'components'`);

      expect(
        cronkPage.isValidConfig({
          id: 'test:page',
          components: [],
        }),
      ).toEqual(false);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'streams'`);

      expect(
        cronkPage.isValidConfig({
          id: 'test:page',
          components: [],
          streams: [],
        }),
      ).toEqual(true);
      expect(cronkPage.stringifyAjvError()).toEqualText('null');
    });
  });

  describe('public methods', () => {
    it('validateConfig() & stringifyAjvError()', async () => {
      page = await newSpecPage({
        components: [CronkPage],
        html: '<cronk-page></cronk-page>',
      });
      const cronkPage = page.rootInstance;
      expect(cronkPage.stringifyAjvError()).toBeUndefined();

      expect(await cronkPage.validateConfig({})).toEqual(false);
      expect((console.error as jest.Mock).mock.calls).toHaveLength(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toContain(`must have required property 'id'`);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'id'`);

      expect(
        await cronkPage.validateConfig({
          id: 'test:page',
        }),
      ).toEqual(false);
      expect((console.error as jest.Mock).mock.calls).toHaveLength(2);
      expect((console.error as jest.Mock).mock.calls[1][0]).toContain(`must have required property 'components'`);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'components'`);

      expect(
        await cronkPage.validateConfig({
          id: 'test:page',
          components: [],
        }),
      ).toEqual(false);
      expect((console.error as jest.Mock).mock.calls).toHaveLength(3);
      expect((console.error as jest.Mock).mock.calls[2][0]).toContain(`must have required property 'streams'`);
      expect(cronkPage.stringifyAjvError()).toContain(`must have required property 'streams'`);

      expect(
        await cronkPage.validateConfig({
          id: 'test:page',
          components: [],
          streams: [],
        }),
      ).toEqual(true);
      expect((console.error as jest.Mock).mock.calls).toHaveLength(3);
      expect(cronkPage.stringifyAjvError()).toEqualText('null');
    });
  });

  describe('rendering', () => {
    it('renders no pageConfig correctly', async () => {
      await page.waitForChanges();
      expect(rootEl).toHaveClass('empty-page-config');
    });

    it('renders various pageConfig correctly', async () => {
      await page.waitForChanges();
      expect(rootEl).toHaveClass('empty-page-config');
      expect(rootEl).not.toHaveClass('valid-page-config');

      rootEl.pageConfig = {} as ReportDefinition;
      await page.waitForChanges();
      expect(rootEl).not.toHaveAttribute('id');
      expect(rootEl).not.toHaveClass('empty-page-config');
      expect(rootEl).not.toHaveClass('valid-page-config');
      expect(rootEl.innerText).toContain(`must have required property 'id'`);

      rootEl.pageConfig = {
        id: 'test:page',
      } as ReportDefinition;
      await page.waitForChanges();

      expect(rootEl).not.toHaveAttribute('id');
      expect(rootEl).not.toHaveClass('empty-page-config');
      expect(rootEl).not.toHaveClass('valid-page-config');

      rootEl.pageConfig = {
        id: 'test:page',
        components: [],
        streams: [],
      };
      await page.waitForChanges();
      expect(rootEl).toHaveAttribute('id');
      expect(rootEl).toEqualAttribute('id', 'test:page');
      expect(rootEl).not.toHaveClass('empty-page-config');
      expect(rootEl).toHaveClass('valid-page-config');
    });
  });
});
