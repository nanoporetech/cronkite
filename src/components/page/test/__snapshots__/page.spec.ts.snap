// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cronk-page rendering renders no pageConfig correctly 1`] = `<cronk-page></cronk-page>`;

exports[`cronk-page rendering renders various pageConfig correctly 1`] = `<cronk-page></cronk-page>`;

exports[`cronk-page rendering renders various pageConfig correctly 2`] = `
<cronk-page><pre style="padding: 0 1rem;">[
  {
    "instancePath": "",
    "schemaPath": "#/required",
    "keyword": "required",
    "params": {
      "missingProperty": "id"
    },
    "message": "must have required property 'id'"
  }
]</pre>
</cronk-page>
`;

exports[`cronk-page rendering renders various pageConfig correctly 3`] = `
<cronk-page><pre style="padding: 0 1rem;">[
  {
    "instancePath": "",
    "schemaPath": "#/required",
    "keyword": "required",
    "params": {
      "missingProperty": "components"
    },
    "message": "must have required property 'components'"
  }
]</pre>
</cronk-page>
`;

exports[`cronk-page rendering renders various pageConfig correctly 4`] = `
<cronk-page class="validate" id="test:page">
  <cronk-page-components id="test:page-components" slot="components"></cronk-page-components>
</cronk-page>
`;
